rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,1053 @@
+model = tf.keras.models.Model(inputs=inputs, outputs=predictions)
+dataset = tf.data.Dataset.zip((features, labels))
+strategy = tf.distribute.MirroredStrategy()
+model.compile(loss='mean_squared_error',
+              optimizer=tf.train.GradientDescentOptimizer(learning_rate=0.2),
+              distribute=strategy)
+model.fit(dataset, epochs=5, steps_per_epoch=10)
+```
+
+As you can see, the only distribution-specific changes for the users are creating the `MirroredStrategy` object, and pass it to the compile call. They don’t need to know / use any other APIs of `MirroredStrategy` - `keras` and other TF components will do that on their behalf.
+
+
+## Design 
+
+In this section, we describe the core DS APIs in detail, with a focus on use case #2 (users writing custom training loops). We will also talk briefly about some of the extended APIs and how to access them.
+
+### Overview
+
+The basic idea is that the user can write their code as if it was running on a single device. By adding just a few lines of DS code, the computation can then be run across multiple machines and / or devices.
+The computation that is running on each device is a [replica](#replica) of the
+user computation (assuming no [model parallelism](#model-parallelism) across
+devices).
+
+The DS API provides a number of primitives to facilitate communication
+across replicas. The most common use case is the aggregation of gradients for
+synchronous SGD training. DS natively supports this by automatically
+adding cross-replica gradient accumulation to any TensorFlow optimizer created
+within a DS scope (see the
+[Simple Classification](#simple-classification) example below). For more
+advanced use cases (such as cross-replica batch normalisation), the API provides
+collective operations for sending and receiving tensors across replicas.
+
+The TF sub-graph generating input data is treated separately from the sub-graph
+performing the main computation (e.g. training a model). This decoupling
+provides flexibility in how input pipelines are connected to replicas. For
+example, replicas running on the same worker can be fed from the same input
+pipeline, which is often more efficient than having a separate input pipeline
+per replica.
+
+The API has been designed to work for all types of devices, including TPUs, so
+that code can be seamlessly moved between different device targets.
+
+### Concepts / Terminology
+
+#### Replica
+
+A replica is "one copy of the computation" that the user wants to replicate. You
+can think of it as "one copy of the model". We might also refer to this as "compute replica" or
+"model replica" in some places.
+
+When using model parallelism, a replica will span several physical devices (e.g.
+GPUs, TPU cores) and may also span multiple physical machines.
+
+`replica_id` will range from 0 to N-1 if N is the number of replicas.
+
+#### Worker
+
+The physical machine(s) containing the physical devices (e.g. GPUs, TPUs) on
```

Comment:
this definition is a bit confusing to me - is a worker one physical machine or something that runs the replica code on one or more physical machines?

Reply:
A worker, as defined here, is the minimum of (one machine, one replica).
So in the most common case, it would correspond to one machine, and has one or multiple replicas.
But in the less common case where one replica needs to span > 1 machine, we are defining a worker to span the minimum number of machines needed for 1 replica.

I agreed this is a little confusing, and I am also worried this might be confusing with the "worker" task as used in TensorFlow in general (Although they correspond to each other in the common cases). 

Do you have suggestions for a different name, or a different concept we should define here?